File: source/js/pages/x_app_analytics.js
export const TITLE = window.Lang.use("app_analytics");

export default async function main(){ return `<container class="max-w-1200px p-5 gap-1"></container>`; }

export async function after(){
	const container = document.querySelector("container");
	Loading.on_element(container);
	container.insertAdjacentHTML("beforeend", await build_HTML());
	Loading.on_element(container);

	async function build_HTML(){
		let resp = await window.bridge({for:"get_all_access_logs"});
		if("data" in resp) resp = resp["data"];
		else return `<p class="w-100 text-size-0-8 surface-info p-1">No data to show.</p>`;

		return `
			<column class="surface-v1 p-2 gap-1 w-100 flex-y-start">

				<p>${resp[0][1]} - ${resp[resp.length-1][1]}</p>

				${build_requests_by_counts_HTML()}

				${build_raw_requests_HTML()}

			</column>
		`;

		function build_requests_by_counts_HTML(){
			// Formatting duplicates
			const requests_by_counts = {};
			for(const cell of resp)
				if(cell[2] in requests_by_counts) requests_by_counts[cell[2]] = requests_by_counts[cell[2]] + 1;
				else requests_by_counts[cell[2]] = 1;

			// List formatting for x-table
			const requests_by_counts_list = [];
			for(const key in requests_by_counts) requests_by_counts_list.push([key, requests_by_counts[key]]);

			return `
				<details class="surface-v1 w-100 text-size-0-8">
					<summary class="surface-v1 bg-4 p-1 px-2">Requests by counts</summary>

					<column class="p-3">
						<x-table class="table-v1 table-zebra table-hover table-bordered w-100">
							{
								"searchable": true,
								"head": [
									{"title": "HTTP method and path", "sortable": true},
									{"title": "Request count", "sortable": true}
								],
								"body": ${JSON.stringify(requests_by_counts_list)}
							}
						</x-table>
					</column>
				</details>
			`;
		}

		function build_raw_requests_HTML(){
			return `
				<details class="surface-v1 w-100 text-size-0-8">
					<summary class="surface-v1 bg-4 p-1 px-2">All requests</summary>

					<column class="p-3">
						<x-table class="table-v1 table-zebra table-hover table-bordered w-100">
							{
								"searchable": true,
								"head": [
									{"title": "IP address"},
									{"title": "Timestamp"},
									{"title": "HTTP method and path"},
									{"title": "Status code"},
									{"title": "Bytes sent"},
									{"title": "Referrer"},
									{"title": "User agent"},
									{"title": "Client IP"},
									{"title": "Response time"}
								],
								"body": ${JSON.stringify(resp)}
							}
						</x-table>
					</column>
				</details>
			`;
		}
	}
}

File: source/python/pages/x_app_analytics.py
import os
import re

from python.modules.Page import Page
from python.modules.response import response
from python.modules.Globals import Globals
from python.modules.Logger import Log

@Page.build()
def x_app_analytics(request):
	if request.method == "POST":
		if request.content_type == "application/json":
			if request.get_json()["for"] == "get_all_access_logs":
				if "URL" not in Globals.CONF["default"]: return response(type="error", message="invalid_request")

				log_file_and_path = f"/var/log/www.{Globals.CONF['default']['URL']['domain_name']}.{Globals.CONF['default']['URL']['domain_extension']}.access.log"

				if not os.path.exists(log_file_and_path):
					Log.warning(f"x_app_analytics -> File does not exists: {log_file_and_path}")
					return response(type="error", message="unknown_error")

				all_lines = []

				try:
					with open(log_file_and_path, 'r') as file:
						for line in file:
							parsed_line = parse_log_line(line.strip())
							if parsed_line is not None: all_lines.append(parsed_line)

				except Exception as e:
					Log.error(f"x_app_analytics -> {e}")
					return response(type="error", message="unknown_error")


				return response(type="success", message="success", data=all_lines, default_serializer_func=str)

def parse_log_line(line):
	pattern = r'(\S+) - - \[(.*?)\] "(.*?)" (\d+) (\d+) "(.*?)" "(.*?)" "(.*?)" response-time=([\d.]+)'
	match = re.match(pattern, line)
	if match:
		return [
			match.group(1), # IP address
			match.group(2), # Timestamp
			match.group(3), # HTTP method and path
			match.group(4), # Status code
			match.group(5), # Bytes sent
			match.group(6), # Referrer
			match.group(7), # User agent
			match.group(8), # Client IP (repeated)
			match.group(9)  # Response time
		]
	return None
